walnuts_cpp/nuts.hpp

```diff
@@ -300,25 +240,39 @@ void transition(Random<S, Generator>& rng,
       if (uturn_flag) break;
     }
   }
-  theta_next = span_accum.theta_select_;
+  theta_next = std::move(span_accum.theta_select_);
 }
 
-template <typename S, class F, class Generator>
+template <typename S, class F, class Generator, class H>
 void nuts(Generator& generator,
           const F& logp_grad_fun,
           const Vec<S>& inv_mass,
           S step,
           Integer max_depth,
-          const Vec<S>& theta,
-          Matrix<S>& sample) {
+          const Vec<S>& theta_init,
+          Integer num_draws,
+          H& handler) {
   Random<S, Generator> rng{generator};
-  Integer num_draws = sample.cols();
-  if (num_draws == 0) return;
-  sample.col(0) = theta;
+  Vec<S> theta = theta_init;  // copy once
+  handler(0, theta);
   for (Integer n = 1; n < num_draws; ++n) {
     transition(rng, logp_grad_fun, inv_mass, step, max_depth,
-               Vec<S>(sample.col(n - 1)), sample.col(n));
+               std::move(theta), theta);
```

Comment:
This doesn’t look right to me, as it’s using “theta” after the call to move. I think the signature of transition needs to change and/or these need to be two separate variables.

Reply:
My intention was to have the call to `transition()` move the data out of `theta`, then have it write the result back into `theta` for the return.  

Isn't this the case you said is guaranteed to work?

Reply:
I think it is valid, but it sets off a lot of red flags to see it. In particular, if the code in transition ever does anything other than move into it, I think it would instantly be UB, or at the very least a hard to track down bug. Looking at it now, I think I would have the transition function return the next theta (which you can instantly move into theta), that way a “dangerous” reference like this never needs to exist.

Reply:
Thanks.  Did you mean a call that would look like the following?

```cpp
theta = transition(rng, logp_grad_fun, inv_mass, step, max_depth,
               Vec<S>(sample.col(n - 1)), sample.col(n));
               std::move(theta));
```

I think this would also work, but doesn't it wave a morally equivalent red flag?

Reply:
^I like that more

Reply:
I also prefer that. To me the difference is that there is no moment in time where an invalid variable exists when written like that

Reply:
I'll change it, but I'd like to sit down in person and have you explain to me what you don't like about the first version that the second version solves.  I didn't think there would ever be a point where `theta` is in an invalid state---I thought the move would reset it to size zero.  And I don't see how if it's in an invalid state in the first call it's not in an equally invalid state at some point in the second call.

Reply:
Fixed.  Move theta into a moved return.